src/modules/sensors/camera.c

```diff
@@ -290,18 +309,55 @@ static void recv_frame(int i) {
     state.brightness[i] = compute_brightness(buf.bytesused) / CAMERA_ILL_MAX;
 }
 
+static void reset_histogram(void) {
+    state.hist.count = memset(state.hist.count, 0, HISTOGRAM_STEPS * sizeof(double));
+    state.hist.sum = memset(state.hist.sum, 0, HISTOGRAM_STEPS * sizeof(double));
+}
+
 static double compute_brightness(const unsigned int size) {
     double brightness = 0.0;
+    double min = CAMERA_ILL_MAX;
+    double max = 0.0;
+
     /*
      * If greyscale -> increment by 1. 
      * If YUYV -> increment by 2: we only want Y! 
      */
     const int inc = 1 + (state.pixelformat == V4L2_PIX_FMT_YUYV);
-    
+
+    // Find minimum and maximum brightness
+    for (int i = 0; i < size; i += inc) {
+        if (state.buf.start[i] < min) {
+            min = state.buf.start[i];
+        }
+        if (state.buf.start[i] > max) {
+            max = state.buf.start[i];
+        }
+    }
+    if (max == 0.0) {
+        return brightness;
+    }
+
+    // Calculate histogram
+    const double step_size = (max - min) / HISTOGRAM_STEPS;
     for (int i = 0; i < size; i += inc) {
-        brightness += state.buf.start[i];
+        int bucket = (state.buf.start[i] - min) / step_size;
+        if (bucket >= 0 && bucket < HISTOGRAM_STEPS) {
+            state.hist.sum[bucket] += state.buf.start[i];
+            state.hist.count[bucket]++;
+        }
     }
-    brightness /= state.width * state.height;
+
+    // Default to the average for the highest brightness bucket
+    brightness = state.hist.sum[HISTOGRAM_STEPS-1] / state.hist.count[HISTOGRAM_STEPS-1];
```

Comment:
Is this really needed though? We parse it anyway!

Reply:
I _may_ be able to remove this - I added it as a safety-net when I was testing out the filter mechanism below, in case we didn't match any buckets.

Reply:
It makes sense, thanks!

Reply:
But are we sure that
`state.hist.count[HISTOGRAM_STEPS-1] != 0`
in any case?

Reply:
Since we've calculated the min/max, the upper bucket is guaranteed to have samples.

Reply:
You're right, thanks for the clarification!